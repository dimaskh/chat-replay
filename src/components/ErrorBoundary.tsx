// Libs
import React from 'react'

// Styles
import styled from 'styled-components'

export default class ErrorBoundary extends React.PureComponent {
  public state = {
    error: null,
    info: null,
  }

  public componentDidCatch(error: any, info: any) {
    this.setState({ error, info })
  }

  public render() {
    const { error } = this.state

    if (error) {
      return <ErrorBoundaryFallbackComponent />
    }

    return this.props.children
  }
}

const Container = styled.div`
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
`

const Message = styled.div`
  padding: 40px;
  border: 2px #78909c solid;
  border-radius: 5px;
  font-size: 24px;
  color: #78909c;
`

const ErrorBoundaryFallbackComponent = () => (
  <Container>
    <Message>
      Some Error Occurred
      <span role="img" aria-label="face-emoji">
        ðŸ˜ž
      </span>
    </Message>
  </Container>
)
