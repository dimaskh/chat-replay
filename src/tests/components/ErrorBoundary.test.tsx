// Libs
import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'

// Tested components
import ErrorBoundary, { ErrorBoundaryFallbackComponent } from 'components/ErrorBoundary'

describe('When no errors are caught in a child component', () => {
  let ErrorBoundaryComponent: any

  beforeAll(() => {
    ErrorBoundaryComponent = shallow(
      <ErrorBoundary>
        <h1>child</h1>
      </ErrorBoundary>
    )
  })

  it('should render the child component', () => {
    expect(ErrorBoundaryComponent.find('h1').exists()).toBeTruthy()
  })
})

describe('When error is caught in a child component', () => {
  let ErrorBoundaryComponent: any

  beforeAll(() => {
    jest.spyOn(global.console, 'log')
    ErrorBoundaryComponent = shallow(
      <ErrorBoundary>
        <h1>child</h1>
      </ErrorBoundary>
    )
    ErrorBoundaryComponent.instance().componentDidCatch('no errors')
    ErrorBoundaryComponent.update()
  })

  it('should log an error to console', () => {
    expect(global.console.log).toHaveBeenCalledWith('no errors')
  })

  it('should update the state to indicate an error', () => {
    expect(ErrorBoundaryComponent.instance().state.error).toBeTruthy()
  })

  it('should not render the child component', () => {
    expect(ErrorBoundaryComponent.find('h1').exists()).toBeFalsy()
  })

  describe('ErrorBoundaryFallbackComponent component', () => {
    it('should render correctly', () => {
      const component = shallow(<ErrorBoundaryFallbackComponent />)

      expect(shallowToJson(component)).toMatchSnapshot()
    })
  })
})
