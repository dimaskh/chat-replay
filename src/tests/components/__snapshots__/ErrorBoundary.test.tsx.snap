// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When error is caught in a child component ErrorBoundaryFallbackComponent component should render correctly 1`] = `
<styled.div>
  <styled.div>
    Some Error Occurred
    <span
      aria-label="face-emoji"
      role="img"
    >
      ðŸ˜ž
    </span>
  </styled.div>
</styled.div>
`;
